rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,412 @@
+package pipeline
+
+import (
+	"errors"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core"
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/core/state"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/metrics"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/circuitcapacitychecker"
+	"github.com/scroll-tech/go-ethereum/rollup/tracing"
+)
+
+type ErrorWithTrace struct {
+	Trace *types.BlockTrace
+	err   error
+}
+
+func (e *ErrorWithTrace) Error() string {
+	return e.err.Error()
+}
+
+func (e *ErrorWithTrace) Unwrap() error {
+	return e.err
+}
+
+var (
+	ErrApplyStageDone           = errors.New("apply stage is done")
+	ErrUnexpectedL1MessageIndex = errors.New("unexpected L1 message index")
+
+	lifetimeTimer   = metrics.NewRegisteredTimer("pipeline/lifetime", nil)
+	applyTimer      = metrics.NewRegisteredTimer("pipeline/apply", nil)
+	applyIdleTimer  = metrics.NewRegisteredTimer("pipeline/apply_idle", nil)
+	applyStallTimer = metrics.NewRegisteredTimer("pipeline/apply_stall", nil)
+	cccTimer        = metrics.NewRegisteredTimer("pipeline/ccc", nil)
+	cccIdleTimer    = metrics.NewRegisteredTimer("pipeline/ccc_idle", nil)
+)
+
+type Pipeline struct {
+	chain    *core.BlockChain
+	vmConfig *vm.Config
+	coinbase *common.Address
+	parent   *types.Block
+	start    time.Time
+
+	// accumalators
+	ccc            *circuitcapacitychecker.CircuitCapacityChecker
+	Header         types.Header
+	state          *state.StateDB
+	nextL1MsgIndex uint64
+	blockSize      common.StorageSize
+	txs            types.Transactions
+	coalescedLogs  []*types.Log
+	receipts       types.Receipts
+	gasPool        *core.GasPool
+
+	// com channels
+	TxnQueue         chan *types.Transaction
+	applyStageRespCh chan error
+	ResultCh         chan *Result
+
+	// Test hooks
+	beforeTxHook func() // Method to call before processing a transaction.
+}
+
+func NewPipeline(
+	chain *core.BlockChain,
+	vmConfig *vm.Config,
+	state *state.StateDB,
+	coinbase *common.Address,
+
+	header *types.Header,
+	ccc *circuitcapacitychecker.CircuitCapacityChecker,
+) *Pipeline {
+	parent := chain.GetBlock(header.ParentHash, header.Number.Uint64()-1)
+	return &Pipeline{
+		chain:          chain,
+		vmConfig:       vmConfig,
+		parent:         parent,
+		nextL1MsgIndex: *rawdb.ReadFirstQueueIndexNotInL2Block(chain.Database(), parent.Hash()),
+		Header:         *header,
+		ccc:            ccc,
+		coinbase:       coinbase,
```

Comment:
This is not really used (see `NewEVMBlockContext`). I think it was only kept for compatibility with old tests. So we can just pass `nil` to `ApplyTransaction ` and add a comment about `FeeVaultEnabled`.

Reply:
removed in https://github.com/scroll-tech/go-ethereum/pull/735/commits/ad086f9b7dcbda95565d91f3d6943dbe07760c50. removing it from worker required to touch a lot more files.